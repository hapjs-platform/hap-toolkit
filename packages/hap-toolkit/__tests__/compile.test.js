/*
 * Copyright (c) 2021-present, the hapjs-platform Project Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

'use strict'

const fs = require('fs')
const path = require('path')
const { Writable } = require('stream')
const del = require('del')
const glob = require('glob')
const stripAnsi = require('strip-ansi')
const { copyApp, wipeDynamic } = require('hap-dev-utils')
const { compile, stopWatch } = require('../lib')
const JSZip = require('jszip')

describe('测试compile', () => {
  const platform = 'native'
  let projectRoot
  let buildDir
  let outputs
  let outputStream
  let sitemapPath
  let babelConfigPath

  beforeAll(async () => {
    const testAppDir = path.resolve(__dirname, '../fixtures/app')
    projectRoot = await copyApp(testAppDir)
    buildDir = path.resolve(projectRoot, 'build')
    outputs = []
    outputStream = new Writable({
      write(chunk, encoding, next) {
        outputs.push(chunk.toString())
        next()
      }
    })
    sitemapPath = path.join(projectRoot, 'src/sitemap.json')
    babelConfigPath = path.join(projectRoot, 'babel.config.js')
  })

  it('generate webpack config', async () => {
    const genWebpackConf = require('../lib/gen-webpack-conf').default
    const conf = await genWebpackConf(
      {
        cwd: projectRoot
      },
      'development'
    )
    expect(conf.entry).toMatchSnapshot()
  })

  it(
    'compile release',
    async () => {
      const mode = 'prod'
      const expectResult = [
        'app.js',
        'CardDemo/index.js',
        'Common/logo.png',
        'Common/tu.9.png',
        'Demo/index.js',
        'workers/request/index.js'
      ]

      const { stats } = await compile(platform, mode, false, {
        cwd: projectRoot
      })
      expect(stats.hasErrors()).toBe(false)

      const result = glob.sync('**/*.{json,css,png,js,js.map}', {
        cwd: buildDir
      })
      expect(result).toEqual(expect.arrayContaining(expectResult))
      expect(result).toMatchSnapshot()
    },
    5 * 60 * 1000
  )

  it(
    'compile preview package',
    async () => {
      // 为保证测试简单，随机写死一个hash
      const userHash = '49ba59abbe56e057'
      const previewPackageSuffix = '.__preview__'
      const mode = 'prod'
      const tempPreviewPath = path.resolve(__dirname, '../fixtures/')
      const manifestPath = path.resolve(buildDir, './manifest.json')
      const { stats } = await compile(platform, mode, false, {
        cwd: projectRoot,
        buildPreviewRpkOptions: {
          setPreviewPkgPath: tempPreviewPath,
          userNameHash: userHash
        }
      })
      expect(stats.hasErrors()).toBe(false)
      // 1、是否正确输出到了指定目录
      const manifest = JSON.parse(fs.readFileSync(manifestPath).toString())
      const previewRpkOutput = `${tempPreviewPath}/${manifest.package}.__preview__.debug.${manifest.versionName}.rpk`
      const err = fs.accessSync(previewRpkOutput, fs.constants.F_OK)
      expect(err).toBe(undefined)
      // 2、输出文件内的包名和name是否被正确替换
      const outputRpkBuf = fs.readFileSync(previewRpkOutput)
      const outputRpkZip = await JSZip.loadAsync(outputRpkBuf)
      const manifestJSON = await outputRpkZip.file('manifest.json').async('string')
      const pkgNameExistResult =
        JSON.parse(manifestJSON).package.indexOf(previewPackageSuffix) !== -1
      const nameExistResult = JSON.parse(manifestJSON).name === userHash
      expect(pkgNameExistResult).toBe(true)
      expect(nameExistResult).toBe(true)
      // 3、原build目录内的包名是否正确被重置
      const manifestJsonOther = JSON.parse(fs.readFileSync(manifestPath).toString())
      const pkgNameResetResult = manifestJsonOther.package.indexOf(previewPackageSuffix) === -1 // 说明包名被重置了
      const nameResetResult = manifestJsonOther.name !== userHash // 说明name被重置了
      expect(pkgNameResetResult).toBe(true)
      expect(nameResetResult).toBe(true)
      // 删除临时rpk文件
      const _err = fs.unlinkSync(previewRpkOutput)
      expect(_err).toBe(undefined)
    },
    5 * 60 * 1000
  )

  // 这里会记录很多内容
  it(
    'compile build devtool=false',
    async () => {
      const mode = 'dev'
      const expectResult = [
        'app.js',
        'CardDemo/index.js',
        'Demo/index.js',
        'workers/request/index.js'
      ]

      // 第三个参数为是否开启watch，true为开启
      const data = await compile(platform, mode, true, {
        cwd: projectRoot,
        log: outputStream,
        devtool: false
      })
      expect(data.compileError).toBeNull()
      expect(data.stats.hasErrors()).toBe(false)

      const result = glob.sync('**/*.{js,js.map}', {
        cwd: buildDir
      })
      expect(result).toEqual(expect.arrayContaining(expectResult))
      expect(result).toMatchSnapshot()

      // 更详细的 snapshots
      const json = data.stats.toJson({
        source: true
      })
      json.modules.forEach((module, index) => {
        expect(wipeDynamic(module.source, [[projectRoot, '<project-root>']])).toMatchSnapshot()
      })
      expect(json.assets.map((a) => a.name)).toMatchSnapshot('assets list')
      // eg. '\u001B[4mUnicorn\u001B[0m' => 'Unicorn'
      const output = stripAnsi(outputs.join('\n'))
      expect(wipeDynamic(output)).toMatchSnapshot('outputs')
    },
    5 * 60 * 1000
  )

  // 保证上一个 watcher 实例关闭，这里不加就报错
  it('stop watch', async () => {
    const data = await stopWatch()
    expect(data.stopWatchError).toBeNull()
  })

  it(
    'compile build devtool=source-map',
    async () => {
      const mode = 'dev'
      const expectResult = [
        'app.js',
        'app.js.map',
        'CardDemo/index.js',
        'CardDemo/index.js.map',
        'Common/logo.png',
        'Common/tu.9.png',
        'Demo/index.js',
        'Demo/index.js.map',
        'workers/request/index.js',
        'workers/request/index.js.map'
      ]

      // 第三个参数为是否开启watch，true为开启
      const data = await compile(platform, mode, true, { cwd: projectRoot, devtool: 'source-map' })
      expect(data.compileError).toBeNull()
      expect(data.stats.hasErrors()).toBe(false)

      const result = glob.sync('**/*.{json,css,png,js,js.map}', {
        cwd: buildDir
      })
      expect(result).toEqual(expect.arrayContaining(expectResult))
      expect(result).toMatchSnapshot()
    },
    5 * 60 * 1000
  )

  it('stop watch', async () => {
    const data = await stopWatch()
    expect(data.stopWatchError).toBeNull()
  })

  it(
    'validate sitemap',
    async () => {
      const mode = 'dev'
      const errPage = '123'
      const sitemap = {
        doc: 'https://doc.quickapp.cn/framework/sitemap.html',
        rules: [{ rule: 'enable', page: errPage }]
      }
      fs.writeFileSync(sitemapPath, JSON.stringify(sitemap))

      // 第三个参数为是否开启watch，true为开启
      try {
        await compile(platform, mode, false, { cwd: projectRoot, devtool: 'source-map' })
      } catch (err) {
        expect(err.message).toMatch(`sitemap rules第1项配置错误，该page不存在：${errPage}`)
      }
    },
    5 * 60 * 1000
  )

  it(
    'compile without babel.config.js',
    async () => {
      const mode = 'dev'
      // 删去 sitemap 以防校验影响
      fs.unlinkSync(sitemapPath)
      fs.unlinkSync(babelConfigPath)

      const { stats } = await compile(platform, mode, false, {
        cwd: projectRoot
      })
      expect(stats.hasErrors()).toBe(false)
    },
    5 * 60 * 1000
  )

  it(
    'compile with custom babel.config.js plugin',
    async () => {
      const mode = 'dev'
      const customBabelPlugin = '@babel/plugin-custom-demo'
      const babelConfig = `
      module.exports = function(api) {
        api.cache(true)
        return {
          presets: ['@babel/preset-env'],
          plugins: ['@babel/plugin-transform-modules-commonjs', '${customBabelPlugin}'],
          babelrcRoots: ['.', 'node_modules']
        }
      }
      `
      fs.writeFileSync(babelConfigPath, babelConfig)

      const { stats } = await compile(platform, mode, false, {
        cwd: projectRoot
      })
      stats.compilation.errors.forEach((error) => {
        expect(error.message).toMatch(`Cannot find module '${customBabelPlugin}'`)
      })
    },
    5 * 60 * 1000
  )

  afterAll(async () => {
    await del([projectRoot])
  })
})
