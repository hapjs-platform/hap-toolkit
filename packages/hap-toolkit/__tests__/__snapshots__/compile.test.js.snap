// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`测试compile compile build devtool=false 1`] = `
Array [
  "app.js",
  "CardDemo/index.js",
  "Demo/index.js",
  "workers/request/index.js",
]
`;

exports[`测试compile compile build devtool=false 2`] = `
"
var $app_style$ = {}
var $app_script$ = require(\\"!../../../../hap-dsl-xvm/lib/loaders/script-loader.js!../../../../hap-packager/lib/loaders/module-loader.js!../../../../hap-dsl-xvm/lib/loaders/manifest-loader.js?path=<project-root>/src!../../../../../node_modules/babel-loader/lib/index.js?cwd=<project-root>&cacheDirectory&comments=false&configFile=<project-root>/babel.config.js!../../../../hap-dsl-xvm/lib/loaders/fragment-loader.js?index=0&type=script!./app.ux?uxType=app\\")

$app_define$('@app-application/app', [], function ($app_require$, $app_exports$, $app_module$) {
  
  $app_script$($app_module$, $app_exports$, $app_require$)
  if ($app_exports$.__esModule && $app_exports$.default) {
    $app_module$.exports = $app_exports$.default
  }
  $app_module$.exports.manifest = require(\\"./manifest.json\\")
  $app_module$.exports.style = { list: [ $app_style$ ] }
  
})
$app_bootstrap$('@app-application/app', { packagerVersion: QUICKAPP_TOOLKIT_VERSION })
"
`;

exports[`测试compile compile build devtool=false 3`] = `
"
var $app_style$ = require(\\"!../../../../../hap-dsl-xvm/lib/loaders/style-loader.js?index=0&type=style!../../../../../hap-dsl-xvm/lib/loaders/fragment-loader.js?index=0&type=style!./index.ux?uxType=page\\")

var $app_script$ = require(\\"!../../../../../hap-dsl-xvm/lib/loaders/script-loader.js!../../../../../hap-packager/lib/loaders/module-loader.js!../../../../../../node_modules/babel-loader/lib/index.js?cwd=<project-root>&cacheDirectory&plugins[]=<CWD>/packages/hap-dsl-xvm/lib/loaders/babel-plugin-jsx.js&comments=false&configFile=<project-root>/babel.config.js!../../../../../hap-dsl-xvm/lib/loaders/access-loader.js!../../../../../hap-dsl-xvm/lib/loaders/fragment-loader.js?index=0&type=script!./index.ux?uxType=page\\")

$app_define$('@app-component/index', [], function($app_require$, $app_exports$, $app_module$) {
     $app_script$($app_module$, $app_exports$, $app_require$)
         if ($app_exports$.__esModule && $app_exports$.default) {
            $app_module$.exports = $app_exports$.default
        }
     $app_module$.exports.template = require(\\"!../../../../../hap-dsl-xvm/lib/loaders/template-loader.js!../../../../../hap-dsl-xvm/lib/loaders/fragment-loader.js?index=0&type=template!./index.ux?uxType=page&\\")

     $app_module$.exports.style = $app_style$
})
$app_bootstrap$('@app-component/index',{ packagerVersion: QUICKAPP_TOOLKIT_VERSION })"
`;

exports[`测试compile compile build devtool=false 4`] = `
"
var $app_style$ = require(\\"!../../../../../hap-dsl-xvm/lib/loaders/style-loader.js?index=0&type=style!../../../../../hap-dsl-xvm/lib/loaders/fragment-loader.js?index=0&type=style!./index.ux?uxType=card\\")

var $app_script$ = require(\\"!../../../../../hap-dsl-xvm/lib/loaders/script-loader.js!../../../../../hap-packager/lib/loaders/module-loader.js!../../../../../../node_modules/babel-loader/lib/index.js?cwd=<project-root>&cacheDirectory&plugins[]=<CWD>/packages/hap-dsl-xvm/lib/loaders/babel-plugin-jsx.js&comments=false&configFile=<project-root>/babel.config.js!../../../../../hap-dsl-xvm/lib/loaders/access-loader.js!../../../../../hap-dsl-xvm/lib/loaders/fragment-loader.js?index=0&type=script!./index.ux?uxType=card\\")

$app_define$('@app-component/index', [], function($app_require$, $app_exports$, $app_module$) {
     $app_script$($app_module$, $app_exports$, $app_require$)
         if ($app_exports$.__esModule && $app_exports$.default) {
            $app_module$.exports = $app_exports$.default
        }
     $app_module$.exports.template = require(\\"!../../../../../hap-dsl-xvm/lib/loaders/template-loader.js!../../../../../hap-dsl-xvm/lib/loaders/fragment-loader.js?index=0&type=template!./index.ux?uxType=card&\\")

     $app_module$.exports.style = $app_style$
})
$app_bootstrap$('@app-component/index',{ packagerVersion: QUICKAPP_TOOLKIT_VERSION })"
`;

exports[`测试compile compile build devtool=false 5`] = `
"\\"use strict\\";

/*
 * Copyright (c) 2021-present, the hapjs-platform Project Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

console.log('hello worker');
postMessage({
  msg: 'hello from worker: ' + utils.test(),
  buffer: utils.str2ab('hello arrayBuffer from worker')
});
onMessage(msg => {
  console.log('[Worker] on appservice message', msg);
  const buffer = msg.buffer;
  console.log('[Worker] on appservice buffer length ', buffer);
  console.log('[Worker] on appservice buffer', utils.ab2str(buffer));
});"
`;

exports[`测试compile compile build devtool=false 6`] = `
"module.exports = function __scriptModule__ (module, exports, $app_require$){\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.default = void 0;
var _default = {};
exports.default = _default;}"
`;

exports[`测试compile compile build devtool=false 7`] = `
"{
  \\"package\\": \\"com.application.demo\\",
  \\"name\\": \\"app\\",
  \\"versionName\\": \\"1.0.0\\",
  \\"versionCode\\": \\"1\\",
  \\"minPlatformVersion\\": 1070,
  \\"icon\\": \\"/Common/logo.png\\",
  \\"features\\": [],
  \\"permissions\\": [{ \\"origin\\": \\"*\\" }],
  \\"config\\": {
    \\"logLevel\\": \\"off\\"
  },
  \\"router\\": {
    \\"entry\\": \\"Demo\\",
    \\"pages\\": {
      \\"Demo\\": {
        \\"component\\": \\"index\\"
      }
    },
    \\"widgets\\": {
      \\"CardDemo\\": {
        \\"name\\": \\"CardDemo\\",
        \\"decription\\": \\"快应用卡片展示\\",
        \\"component\\": \\"index\\",
        \\"path\\": \\"/CardDemo\\",
        \\"features\\": []
      }
    }
  },
  \\"workers\\": {
    \\"entries\\": [
      {
        \\"file\\": \\"workers/request/index.js\\"
      }
    ]
  }
}
"
`;

exports[`测试compile compile build devtool=false 8`] = `
"module.exports = {
  \\".demo-page\\": {
    \\"flexDirection\\": \\"column\\",
    \\"justifyContent\\": \\"center\\",
    \\"alignItems\\": \\"center\\"
  },
  \\".title\\": {
    \\"fontSize\\": \\"40px\\",
    \\"textAlign\\": \\"center\\"
  },
  \\".btn\\": {
    \\"width\\": \\"550px\\",
    \\"height\\": \\"86px\\",
    \\"marginTop\\": \\"75px\\",
    \\"borderRadius\\": \\"43px\\",
    \\"backgroundColor\\": \\"#09ba07\\",
    \\"fontSize\\": \\"30px\\",
    \\"color\\": \\"#ffffff\\"
  }
}"
`;

exports[`测试compile compile build devtool=false 9`] = `
"module.exports = function __scriptModule__ (module, exports, $app_require$){\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.default = void 0;
var _default = {
  private: {
    title: '示例页面'
  }
};
exports.default = _default;
const moduleOwn = exports.default || module.exports;
const accessors = ['public', 'protected', 'private'];
if (moduleOwn.data && accessors.some(function (acc) {
  return moduleOwn[acc];
})) {
  throw new Error('页面VM对象中的属性data不可与\\"' + accessors.join(',') + '\\"同时存在，请使用private替换data名称');
} else if (!moduleOwn.data) {
  moduleOwn.data = {};
  moduleOwn._descriptor = {};
  accessors.forEach(function (acc) {
    const accType = typeof moduleOwn[acc];
    if (accType === 'object') {
      moduleOwn.data = Object.assign(moduleOwn.data, moduleOwn[acc]);
      for (const name in moduleOwn[acc]) {
        moduleOwn._descriptor[name] = {
          access: acc
        };
      }
    } else if (accType === 'function') {
      console.warn('页面VM对象中的属性' + acc + '的值不能是函数，请使用对象');
    }
  });
}}"
`;

exports[`测试compile compile build devtool=false 10`] = `
"module.exports = {
  \\"type\\": \\"div\\",
  \\"attr\\": {},
  \\"classList\\": [
    \\"demo-page\\"
  ],
  \\"children\\": [
    {
      \\"type\\": \\"text\\",
      \\"attr\\": {
        \\"value\\": function () {return '' + '欢迎打开' + (this.title)}
      },
      \\"classList\\": [
        \\"title\\"
      ]
    }
  ]
}"
`;

exports[`测试compile compile build devtool=false 11`] = `
"module.exports = {
  \\".demo-page\\": {
    \\"flex\\": 1,
    \\"flexDirection\\": \\"column\\",
    \\"alignItems\\": \\"center\\"
  }
}"
`;

exports[`测试compile compile build devtool=false 12`] = `
"module.exports = function __scriptModule__ (module, exports, $app_require$){\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.default = void 0;
var _default = {
  private: {
    text: '欢迎使用快应用卡片',
    des: '卡片是一种轻量级的快应用，快应用卡片使快应用能够在其他app和系统里提供可扩展的app功能'
  }
};
exports.default = _default;
const moduleOwn = exports.default || module.exports;
const accessors = ['public', 'protected', 'private'];
if (moduleOwn.data && accessors.some(function (acc) {
  return moduleOwn[acc];
})) {
  throw new Error('页面VM对象中的属性data不可与\\"' + accessors.join(',') + '\\"同时存在，请使用private替换data名称');
} else if (!moduleOwn.data) {
  moduleOwn.data = {};
  moduleOwn._descriptor = {};
  accessors.forEach(function (acc) {
    const accType = typeof moduleOwn[acc];
    if (accType === 'object') {
      moduleOwn.data = Object.assign(moduleOwn.data, moduleOwn[acc]);
      for (const name in moduleOwn[acc]) {
        moduleOwn._descriptor[name] = {
          access: acc
        };
      }
    } else if (accType === 'function') {
      console.warn('页面VM对象中的属性' + acc + '的值不能是函数，请使用对象');
    }
  });
}}"
`;

exports[`测试compile compile build devtool=false 13`] = `
"module.exports = {
  \\"type\\": \\"div\\",
  \\"attr\\": {},
  \\"classList\\": [
    \\"demo-page\\"
  ],
  \\"children\\": [
    {
      \\"type\\": \\"text\\",
      \\"attr\\": {
        \\"value\\": function () {return this.text}
      }
    },
    {
      \\"type\\": \\"text\\",
      \\"attr\\": {
        \\"value\\": function () {return this.des}
      }
    }
  ]
}"
`;

exports[`测试compile compile build devtool=false: assets list 1`] = `
Array [
  "Demo/index.js",
  "CardDemo/index.js",
  "app.js",
  "workers/request/index.js",
]
`;

exports[`测试compile compile build devtool=false: outputs 1`] = `
"[WARN] 您使用了自定义 babel.config.js，请确认是否有安装相关依赖

[ERROR] 校验 manifest.json 错误 
versionCode 字段必须为 Integer 类型
参考: https://doc.quickapp.cn/framework/manifest.html

[LOG] 当前minPlatformVersion >= 1040，平台采用v8版本为6.5+（对应chrome版本为65版+），工具将不再对V8 6.5版本支持的ES6代码进行转换

[INFO] 配置环境：{\\"NODE_PLATFORM\\":\\"native\\",\\"NODE_PHASE\\":\\"dev\\"}

[INFO] 获取到app文件后缀: ux

[INFO] Build Time Cost: : <time-cost>

[LOG] ### App Loader ### 项目构建启用了用户 debug 签名

[LOG] ### App Loader ### 项目构建并生成文件：com.application.demo.__preview__.debug.1.0.0.rpk

[LOG] ### App Server ### 没有记录手机地址，不会通知手机更新rpk文件
"
`;

exports[`测试compile compile build devtool=source-map 1`] = `
Array [
  "app.js",
  "app.js.map",
  "CardDemo/index.js",
  "CardDemo/index.js.map",
  "Common/logo.png",
  "Common/tu.9.png",
  "Demo/index.js",
  "Demo/index.js.map",
  "manifest.json",
  "workers/request/index.js",
  "workers/request/index.js.map",
]
`;

exports[`测试compile compile release 1`] = `
Array [
  "app.js",
  "CardDemo/index.js",
  "Common/logo.png",
  "Common/tu.9.png",
  "Demo/index.js",
  "manifest.json",
  "workers/request/index.js",
]
`;

exports[`测试compile generate webpack config 1`] = `
Object {
  "CardDemo/index": "./src/CardDemo/index.ux?uxType=card",
  "Demo/index": "./src/Demo/index.ux?uxType=page",
  "app": "./src/app.ux?uxType=app",
  "workers/request/index": "./src/workers/request/index.js",
}
`;
